apps/bfd-pipeline/bfd-pipeline-shared-utils/src/main/java/gov/cms/bfd/pipeline/sharedutils/SamhsaUtil.java

```diff
@@ -0,0 +1,392 @@
+      detailsList.add(detail);
+    }
+  }
+
+  /**
+   * Constructs a list of TagDetail objects for a Fiss claim.
+   *
+   * @param fissClaim The claim to check.
+   * @return a list of TagDetail objects, one for each SAMHSA code found in the claim.
+   */
+  private Optional<List<TagDetails>> getPossibleFissSamhsaFields(RdaFissClaim fissClaim) {
+    List<TagDetails> entries = new ArrayList<>();
+    LocalDate serviceDate =
+        fissClaim.getStmtCovFromDate() == null
+            ? LocalDate.parse("1970-01-01")
+            : fissClaim.getStmtCovFromDate();
+    LocalDate throughDate =
+        fissClaim.getStmtCovToDate() == null ? LocalDate.now() : fissClaim.getStmtCovToDate();
+
+    buildDetails(
+        isSamhsaCode(Optional.ofNullable(fissClaim.getAdmitDiagCode())),
+        "fiss_claims",
+        "admit_diag_code",
+        null,
+        entries,
+        serviceDate,
+        throughDate);
+    for (RdaFissRevenueLine revenueLine : fissClaim.getRevenueLines()) {
+      buildDetails(
+          isSamhsaCode(Optional.ofNullable(revenueLine.getApcHcpcsApc())),
+          "fiss_revenue_lines",
+          "apc_hcpcs_apc",
+          (int) revenueLine.getRdaPosition(),
+          entries,
+          serviceDate,
+          throughDate);
+      buildDetails(
+          isSamhsaCode(Optional.ofNullable(revenueLine.getHcpcCd())),
+          "fiss_revenue_lines",
+          "hcpcs_cd",
+          (int) revenueLine.getRdaPosition(),
+          entries,
+          serviceDate,
+          throughDate);
+    }
+    buildDetails(
+        isSamhsaCode(Optional.ofNullable(fissClaim.getDrgCd())),
+        "fiss_claims",
+        "drg_cd",
+        null,
+        entries,
+        serviceDate,
+        throughDate);
+    buildDetails(
+        isSamhsaCode(Optional.ofNullable(fissClaim.getPrincipleDiag())),
+        "fiss_claims",
+        "principle_diag",
+        null,
+        entries,
+        serviceDate,
+        throughDate);
+    for (RdaFissDiagnosisCode diagCode : fissClaim.getDiagCodes()) {
+      buildDetails(
+          isSamhsaCode(Optional.ofNullable(diagCode.getDiagCd2())),
+          "fiss_diagnosis_codes",
+          "diag_cd2",
+          (int) diagCode.getRdaPosition(),
+          entries,
+          serviceDate,
+          throughDate);
+    }
+    for (RdaFissProcCode procCode : fissClaim.getProcCodes()) {
+      buildDetails(
+          isSamhsaCode(Optional.ofNullable(procCode.getProcCode())),
+          "fiss_proc_codes",
+          "proc_code",
+          (int) procCode.getRdaPosition(),
+          entries,
+          serviceDate,
+          throughDate);
+    }
+    return entries.isEmpty() ? Optional.empty() : Optional.of(entries);
+  }
+
+  /**
+   * Check if a given code is a SAMHSA code.
+   *
+   * @param code the code to check.
+   * @return If the code is SAMHSA, returns the SAMHSA entry. Otherwise, an empty optional.
+   */
+  public Optional<SamhsaEntry> isSamhsaCode(Optional<String> code) {
```

Comment:
Maybe this should be called `getSamhsaCode`? `isSamhsaCode` implies that it returns a boolean.